Service accounts (#15599)
DataDog/documentation: content/en/workflows/service_accounts.md

---

> title: Service accounts for Workflows

Author: ```suggestion
title: Service Accounts for Workflows
```

kind: Documentation
disable_toc: false
further_reading:
- link: "/account_management/org_settings/service_accounts/"
  tag: "Documentation"
  text: "Learn more about Datadog service accounts"
---

> To add am automated or scheduled trigger to a workflow, the workflow must have an associated service account. The service account is needed:

Author: ```suggestion
To add an automated or scheduled trigger to a workflow, the workflow must have an associated service account. The service account is needed:
```

- to resolve the connections defined in the workflow actions at runtime
- to provide an identity for workflow executions
- to provide an identity for workflow audit trails

To use automated or scheduled triggers, a workflow must be associated with a unique service account.

## Requirements

To create a service account for a workflow:
- You must have either the Datadog admin role, or a custom role with the `SERVICE_ACCOUNT_WRITE` permission. For more information on service accounts and permissions, see [Service accounts][1].

> - Before you can add a service account, you need a workflow with at least one action so you can add a Trigger.

Author: ```suggestion
- Before you can add a service account, you need a workflow with at least one action so you can add a trigger.
```

## Associate a service account with a workflow

> You can dynamically create a service account for your workflow when you [add a trigger][5]. Alternatively, click on the workflow name add a service account next to **@worklow-** to dynamically create a service account and associate it with the workflow.

Author: ```suggestion
You can dynamically create a service account for your workflow when you [add a trigger][5]. Alternatively, click on the workflow name and add a service account next to **@worklow-** to dynamically create a service account and associate it with the workflow.
```

---
title: Service Accounts for Workflows
kind: Documentation
disable_toc: false
further_reading:
- link: "/account_management/org_settings/service_accounts/"
  tag: "Documentation"
  text: "Learn more about Datadog service accounts"
---

To add an automated or scheduled trigger to a workflow, the workflow must have an associated service account. The service account is needed to:
- resolve the connections defined in the workflow actions at runtime
- provide an identity for workflow executions
- provide an identity for workflow audit trails

> To use automated or scheduled triggers, a workflow must be associated with a unique service account.

Reviewer: Lines 11 and 16 are redundant. Is that intentional?

## Requirements

To create a service account for a workflow:

> - You must have either the Datadog admin role, or a custom role with the `SERVICE_ACCOUNT_WRITE` permission. For more information on service accounts and permissions, see [Service accounts][1].

Reviewer 2: In the roles UI, this is displayed as `Service Account Write` (not `SERVICE_ACCOUNT_WRITE`). https://drive.google.com/file/d/1uja275fpKfLq0hlYOpdyGjw3L2wojqGw/view?usp=sharing

> - Before you can add a service account, you need a workflow with at least one action.

Reviewer 3: This feels like it should preface all the other requirements. (At the very least it doesn't seem like it's related to creating the service account)

## Associate a service account with a workflow

> You can dynamically create a service account for your workflow when you [add a trigger][5]. Alternatively, click on the workflow name and add a service account next to **@worklow-** to dynamically create a service account and associate it with the workflow.

Reviewer 2: I'm not sure I follow the `**@worklow-**` part. I don't see anything like that in the UI when I click the workflow name: https://drive.google.com/file/d/1rP_BN_N9lqK7mMyaxAdnGUvddweYOc_L/view?usp=sharing

{{< img src="workflows/create-service-account.png" alt="You can create a service account by clicking on the workflow name" style="width:70%;" >}}

Service accounts use the email address associated with your Datadog account and follow the pattern: `[WF]:<your-workflow-name>`.

## Service account roles and permissions

> When a workflow action runs, it checks the service account's roles to make sure that it is allowed to execute. For example, the **AWS Lambda** action requires the `CONNECTION_RESOLVE` resource permissions. The Datadog Admin Role and Datadog Standard Role both have `CONNECTION_RESOLVE` resource permissions, so the service account would require one of those roles before the **AWS Lambda** action could execute.

Reviewer: ```suggestion
When a workflow action runs, it checks the service account's roles to make sure that it is allowed to execute. For example, the **AWS Lambda** action requires the `CONNECTION_RESOLVE` resource permission. The Datadog Admin Role and Datadog Standard Role both have the `CONNECTION_RESOLVE` resource permission, so the service account requires one of those roles before the **AWS Lambda** action could execute.
```

Reviewer 2: The `CONNECTION_RESOLVE` permission isn't visible in the UI. It's split from the Datadog Standard Role. (And admins also have it.) What you wrote is consistent with that. But is it confusing to refer to a permission that isn't visible in the UI? And would it be simpler to just say something like: workflow service accounts must have the Datadog Standard Role or Datadog Admin Role.

(I'm fine either way! Just want to make sure you're aware that "CONNECTION_RESOLVE" isn't visible anywhere in the UI.)

> When a workflow action runs, it checks the service account's roles to make sure that it is allowed to execute. For example, the **AWS Lambda** action requires the `CONNECTION_RESOLVE` resource permissions. The Datadog Admin Role and Datadog Standard Role both have `CONNECTION_RESOLVE` resource permissions, so the service account would require one of those roles before the **AWS Lambda** action could execute.

Reviewer 3: I think there are two resources and concepts that are at play and I think they're being mixed here.

Resources:

* Workflows
* Connections (associated with an instance of an action in a workflow)

Service accounts are only needed during automated execution, and at this point the resource they need access to are connections. I'm not sure this section makes the distinction between those resources clear.

> In addition to roles, you can further restrict a role's workflow permissions. You can restrict the permissions as follows:

Reviewer 2: The viewer/resolver/editor restrictions apply to *connections*, not workflows.

Reviewer 2: Actually, on second thought: I'm wondering we should just remove this viewer/resolver/editor section entirely. It's primarily for restricting access by humans to edit/use sensitive connections. cc @allengeorge - what do you think?

> - **Viewer**: Can view workflows

Reviewer: I had to include this same information in the authorization page, and used a description list (https://datadoghq.atlassian.net/wiki/spaces/docs4docs/pages/1182958054/Markdown#Description-lists). Might be a good idea here too for consistency?

Reviewer 3: [puzzled] This set of permissions doesn't look correct _for workflows_ (it looks correct for connections). What resource are we referring to here?

This is the latest resource we have for permissions: https://datadoghq.atlassian.net/wiki/spaces/WP/pages/2523726383/Workflow+Automation+Permissions+Reference

Also happy to discuss off Github.

- **Resolver**: Can resolve and review workflows
- **Editor**: Can edit, resolve and review workflows

In the example below, the connection `raph-test` is restricted to the **Datadog Admin Role** with **Editor** permissions, and the **Datadog Standard Role** with **Resolver** permissions.

{{< img src="workflows/service-account-example.png" alt="A connection with roles and permissions assigned" style="width:70%;" >}}

At runtime, a workflow using this connection would require a service account with the Datadog Standard Role before it could successfully execute.

## View service account details

To view a service account's details:

> 1. On your workflow canvas, click on the automated or scheduled trigger.

Reviewer 2: I think "automated or scheduled trigger" should just be "automated trigger". (The button in the UI says: "Add an Automated Trigger".) Then when you click "Add an Automated Trigger", you're presented with two options: `@-mention` or `Schedule`.

1. Next to the service account, click **View**.

> {{< img src="workflows/service-account-details.png" alt="Service account details" style="width:70%;" >}}

Reviewer: All of the pictures are blurry, but this one is especially so.

Author: Yeah, I took these from the Google doc and they're not great. We might have to ask someone on the workflows team to take better ones or wait until we have access so we can take some ourselves

Reviewer 2: https://dd.slack.com/archives/C03FL5A70EM/p1666037817392969?thread_ts=1665785286.132229&cid=C03FL5A70EM -- let me know if that works!
